pkg/location/type_git.go

```diff
@@ -0,0 +1,123 @@
+package location
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	"github.com/GoogleContainerTools/kpt/internal/util/parse"
+)
+
+type Git struct {
+	// Repo is the git repository the package.
+	// e.g. 'https://github.com/kubernetes/examples.git'
+	Repo string
+
+	// Directory is the sub directory of the git repository.
+	// e.g. 'staging/cockroachdb'
+	Directory string
+
+	// Ref can be a Git branch, tag, or a commit SHA-1.
+	Ref string
+}
+
+var _ Reference = Git{}
+
+type GitLock struct {
+	Git
+
+	// Commit is the SHA-1 for the last fetch of the package.
+	// This is set by kpt for bookkeeping purposes.
+	Commit string
+}
+
+var _ ReferenceLock = GitLock{}
+
+func NewGit(location string, opts ...Option) (Git, error) {
+	return newGit(location, makeOptions(opts...))
+}
+
+func newGit(location string, opt options) (Git, error) {
+	// args[1] is "" for commands that do not require an output path
+	gitTarget, err := parse.GitParseArgs(opt.ctx, []string{location, ""})
+	var zero parse.GitTarget
+	if err != nil || gitTarget == zero {
+		return Git{}, err
+	}
+
+	dir := gitTarget.Directory
+	if strings.HasPrefix(dir, "/") {
```

Comment:
There is probably some filepath method (or set thereof) that will make sure this works on non-Unix (if it is to be used as part of the CLI interface?

Reply:
In this case I believe it's safe to assume this value where it appears in git uri and kptfile fields is always canonically '/'

That said the `type_dir.go` may not be as lucky - looking at how [DIR] args are pre-processed in kpt - a call to `filepath.Clean` should probably be the first step of parsing